File: examples/excel-defined-names.ts
/**
 * 例子（Epic-28）：构造 xlsx 加几个命名范围。
 *
 * 跑法：
 *   bun run examples/excel-defined-names.ts <output.xlsx>
 */

import { SpreadsheetDocument } from "../src/excel/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: excel-defined-names <output.xlsx>\n");
    process.exit(2);
  }

  const doc = SpreadsheetDocument.create();
  doc.addDefinedName("Revenue", "Sheet1!$A$2:$A$10");
  doc.addDefinedName("Expenses", "Sheet1!$B$2:$B$10");
  doc.addDefinedName("Total", "Revenue-Expenses", { hidden: true });
  doc.addDefinedName("LocalRange", "Sheet1!$C$1", { localSheetId: 0 });

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (${doc.listDefinedNames().length} defined names)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/ExcelDefinedNames.cs
// Replica of examples/excel-defined-names.ts
// Creates an xlsx with 4 defined names: Revenue, Expenses, Total (hidden), LocalRange (localSheetId=0).

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Spreadsheet;

namespace CrossSdkVerify.Replicas;

public static class ExcelDefinedNames
{
    public static void Run(string outputPath)
    {
        using var doc = SpreadsheetDocument.Create(outputPath,
            DocumentFormat.OpenXml.SpreadsheetDocumentType.Workbook);

        var wbPart = doc.AddWorkbookPart();
        wbPart.Workbook = new Workbook();

        var wsPart = wbPart.AddNewPart<WorksheetPart>();
        wsPart.Worksheet = new Worksheet(new SheetData());

        var sheets = wbPart.Workbook.AppendChild(new Sheets());
        sheets.Append(new Sheet
        {
            Id = wbPart.GetIdOfPart(wsPart),
            SheetId = 1,
            Name = "Sheet1"
        });

        // Defined names
        var definedNames = new DefinedNames();

        // Revenue
        definedNames.Append(new DefinedName("Sheet1!$A$2:$A$10") { Name = "Revenue" });

        // Expenses
        definedNames.Append(new DefinedName("Sheet1!$B$2:$B$10") { Name = "Expenses" });

        // Total (hidden)
        definedNames.Append(new DefinedName("Revenue-Expenses") { Name = "Total", Hidden = true });

        // LocalRange (localSheetId=0)
        definedNames.Append(new DefinedName("Sheet1!$C$1") { Name = "LocalRange", LocalSheetId = 0 });

        wbPart.Workbook.Append(definedNames);
        wbPart.Workbook.Save();
    }
}
